newclient/client.go

```diff
@@ -0,0 +1,245 @@
+// Housekeeping URL different from the standard one (e.g. http://localhost:4001).
+// This is not recommendend in production.
+func WithHousekeepingURL(housekeepingURL string) clientOption {
+	return func(c *Client) error {
+		housekeeping, err := url.Parse(housekeepingURL)
+		if err != nil {
+			return err
+		}
+		c.housekeepingURL = housekeeping
+		return nil
+	}
+}
+
+// The WithPairingURL function allows to specify an
+// Pairing URL different from the standard one (e.g. http://localhost:4002).
+// This is not recommendend in production.
+func WithPairingURL(pairingURL string) clientOption {
+	return func(c *Client) error {
+		// check that it's a valid URL
+		pairing, err := url.Parse(pairingURL)
+		if err != nil {
+			return err
+		}
+		c.pairingURL = pairing
+		return nil
+	}
+}
+
+// The WithRealmManagementURL function allows to specify an
+// RealmManagement URL different from the standard one (e.g. http://localhost:4003).
+// This is not recommendend in production.
+func WithRealmManagementURL(realmManagementURL string) clientOption {
+	return func(c *Client) error {
+		realmManagement, err := url.Parse(realmManagementURL)
+		if err != nil {
+			return err
+		}
+		c.realmManagementURL = realmManagement
+		return nil
+	}
+}
+
+// The WithBaseURL function allows to specify the Astarte
+// base URL (e.g. api.your-astarte.org)
+func WithBaseURL(baseURL string) clientOption {
+	return func(c *Client) error {
+		base, err := url.Parse(baseURL)
+		if err != nil {
+			return err
+		}
+		c.baseURL = base
+
+		return nil
+	}
+}
+
+// The WithHTTPClient function allows to specify an httpClient
+// with custom options, e.g. different timeout, or skipTLSVerify
+func WithHTTPClient(httpClient *http.Client) clientOption {
+	return func(c *Client) error {
+		c.httpClient = httpClient
+		return nil
+	}
+}
+
+// The WithToken function allows to specify a JWT
+// token that the client will use to interact with Astarte.
+func WithToken(token string) clientOption {
+	return func(c *Client) error {
+		c.token = token
+		return nil
+	}
+}
+
+// The WithUserAgent function allows to specify the User Agent
+// that the client will use when making http requests.
+func WithUserAgent(userAgent string) clientOption {
+	return func(c *Client) error {
+		c.userAgent = userAgent
+		return nil
+	}
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with no expiry.
+// The client will use that token to interact with Astarte.
+func WithPrivateKey[T privateKeyProvider](privateKey T) clientOption {
+	return WithPrivateKeyWithTTL(privateKey, 0)
+}
+
+// The WithPrivateKey function allows to specify a realm private key,
+// used internally to generate a valid JWT token to all Astarte APIs with a specified expiry (in seconds).
+// The client will use that token to interact with Astarte.
+func WithPrivateKeyWithTTL[T privateKeyProvider](privateKey T, ttlSeconds int64) clientOption {
```

Comment:
We tend to use `expiry` for the JWT, not `TTL` (in fact it's also used in the documentation here), so I'd stick with the standard naming to avoid confusion